File: _site/solutions/93_restore_ip_addresses.py
'''
Given a string containing only digits, restore it by returning all possible valid IP address combinations.
Example:
Input: "25525511135"
Output: ["255.255.11.135", "255.255.111.35"]

'''


class Solution:
    def restoreIpAddresses(self, s):
        """
        :type s: str
        :rtype: List[str]
        """
        ans = []
        for i in range(3):
            for j in range(i + 1, i + 4):
                for m in range(j + 1, j + 4):
                    for n in range(m + 1, m + 4):
                        if m + 1 > len(s) - 1: continue
                        a, b, c, d = s[0:i + 1], s[i + 1:j + 1], s[
                            j + 1:m + 1], s[m + 1:]
                        v = all([self.valid_ip(e) for e in [a, b, c, d]])
                        if v:
                            ans.append(a + "." + b + "." + c + "." + d)
        return list(set(ans))

    def valid_ip(self, s):
        return not (len(s) > 1 and s[0] == '0' or int(s) > 255)


print(Solution().restoreIpAddresses("25525511135"))
print(Solution().restoreIpAddresses("010010"))

File: _site/solutions/93_restore_ip_addresses.rb
# Given a string containing only digits, restore it by returning all possible valid IP address combinations.
# Example:
# Input: "25525511135"
# Output: ["255.255.11.135", "255.255.111.35"]

# @param {String} s
# @return {String[]}
def restore_ip_addresses(s)
  res = split_string(s, 4).keep_if { |i| i.size == 4 }.map { |i| i.join('.') }
end

def split_string(s, k)
  return [[]] if s.nil? || s.size < k
  if k == 1
    return [[]] if s.size > 1 && s[0] == '0' || s.to_i > 255
    return [[s]]
  end

  ans = []
  (0..2).each do |i|
    a = s[0..i]
    return ans if a.size > 1 && a[0] == '0' || a.to_i > 255 # no more valid ones
    split_string(s[i + 1..-1], k - 1).each { |res| ans << [a].concat(res) }
  end
  ans
end

s = '25525511135'
s = '010010'
p restore_ip_addresses(s)
